src/sources/postgres.py

```diff
@@ -33,6 +34,7 @@ class PostgresSource(Source[DataFrame]):
     def __init__(self, db_url: str, query_string: str):
         self.query_string = query_string
         self.engine: sqlalchemy.engine.Engine = create_engine(db_url)
+        self.maybe_query_from_file()
```

Comment:
This flow validates itself only in the case of query_string as a file. I don't think the constructor should validate itself. The user should validate if they want to.

Also, I think the method should be private, called something different and incorporate the setting of the `query_string` as well.

So, 

```suggestion
    def __init__(self, db_url: str, query_string_or_file: str | Path):
        self.engine: sqlalchemy.engine.Engine = create_engine(db_url)
        self._set_query_string(query_string_or_file)
```

Reply:
another idea that doesn't involve guessing is the config accepts `query_string` or `query_file`. We can create two class method constructors `PostgresSource.from_string()` and `PostgresSource.from_path()`

Reply:
> This flow validates itself only in the case of query_string as a file. 
Whoopsie. Fixed by calling `self.validate` at the end of `__init__`

>  I don't think the constructor should validate itself.

Respectfully disagree. Classes should build themselves properly and if that involves validating the client's data and aborting on failures, they should do that. We shouldn't leave that to the client. If the user validates only if they want to, they're gonna provide a buggy file, get weird tracebacks and then report issues about that eventually.

Reply:
Compromise? :) 


```python
    def __init__(self, db_url: str, query_string: str):
        self.engine: sqlalchemy.engine.Engine = create_engine(db_url)
        self._set_query_string(query_string)
        self.validate()

    def validate(self) -> bool:
        try:
            # Try to compile the query without executing it
            with self.engine.connect() as connection:
                connection.execute(text("EXPLAIN " + self.query_string))
                return True
        except SQLAlchemyError as e:
            log.error("Invalid SQL query: %s", str(e))
            return False

    def fetch(self) -> DataFrame:
        df = pd.read_sql_query(self.query_string, con=self.engine)
        return _convert_bytea_to_hex(df)

    def is_empty(self, data: DataFrame) -> bool:
        return data.empty

    def _set_query_string(self, query_string: str) -> None:
        self.query_string = query_string

        if self.query_string.lower().endswith(".sql"):
            return self._set_query_string_from_file()

    def _set_query_string_from_file(self) -> None:
        sql_source = Path(self.query_string)
        if not sql_source.is_file() or not sql_source.exists():
            raise RuntimeError(
                "Detected directive to include an sql file, "
                f"but it doesn't exist or isn't a file: {sql_source}"
            )

        with open(sql_source, "r", encoding="utf-8") as _handle:
            sql = _handle.read()
            self.query_string = sql
```

Reply:
Ok, so it makes sense to validate in the constructor, but what we have here is inconsistent. We should have a Validated class which calls validate in __post__init__ and each of the Source and Destination interfaces should inherit it. Like so:

```py
class Validated(ABC):
    """Abstract base class that enforces validation on initialization"""

    def __post_init__(self):
        if not self.validate():
            raise ValueError(f"Invalid configuration for {self.__class__.__name__}")

    @abstractmethod
    def validate(self) -> bool:
        """Validate the configuration"""
```

Then source would become `class Source(Generic[T], Validated):` and we can remove its validate declaration.



In fact, then validate could become a private method because nobody needs to call it anymore.

Now all instances will always be validated instead of us having to remember to do it.